leaf/Src/leaf-physical.c

```diff
@@ -3985,3 +3985,96 @@ void tStiffString_pluckNoUpdate(tStiffString* const mp, Lfloat amp)
 }
 
 
+
+/// Shaker model
+
+void    tShaker_init(tShaker* const pm, LEAF* const leaf)
+{
+    tShaker_initToPool(pm, &leaf->mempool);
+}
+
+void    tShaker_initToPool(tShaker* const pm, tMempool* const mp)
+{
+    _tMempool* m = *mp;
+    _tShaker* p = *pm = (_tShaker*) mpool_alloc(sizeof(_tShaker), m);
+    p->mempool = m;
+
+    LEAF* leaf = p->mempool->leaf;
+    p->rand = leaf->random;
+
+    tBiQuad_initToPool(&p->resonator, mp);
+    tBiQuad_setResonance(&p->resonator, 4500., 0.99, false);
+    tNoise_initToPool(&p->noise, WhiteNoise, mp);
+
+    p->energy = 0.0;
+    p->level = 0.0;
+    p->energyDecay = 0.98;
+    p->soundDecay = 0.95;
+    p->probability = 0.10;
+}
+
+void    tShaker_free(tShaker* const pm)
+{
+    _tShaker* p = *pm;
+    
+    mpool_free((char*)p, p->mempool);
+}
+
+Lfloat   tShaker_tick(tShaker* const pm)
+{
+    _tShaker* p = *pm;
+
+    // exponential decay for system energy
+    p->energy *= p->energyDecay;
+
+    // stochastic particle collision event
+    if(p->rand() < p->probability)
+        p->level += p->energy * p->gain;
+
+    // exponential decay for sound
+    p->level *= p->soundDecay;
+
+    // add noise, filter
+    Lfloat sample    = p->level * tNoise_tick(&p->noise);
+    Lfloat output    = tBiQuad_tick(&p->resonator, sample);
+    
+    return output;
+}
+
+Lfloat     tShaker_excite   (tShaker* const pm, Lfloat energy)
+{ 
+    _tShaker* p = *pm;
+
+     // excite the system
+    p->energy += energy;
+    return p->energy;
+}
+
+void        tShaker_setEnergyDecay   (tShaker *const pm, Lfloat decay) 
+{
+    _tShaker* p = *pm;
+    p->energyDecay = decay;
+}
+
+void        tShaker_setSoundDecay   (tShaker *const pm, Lfloat decay) 
+{
+    _tShaker* p = *pm;
+    p->soundDecay = decay;
+}
+
+void        tShaker_setProbability (tShaker *const pm, Lfloat prob) 
+{
+    _tShaker* p = *pm;
+
+    // set gain correction for number of particles
+    float nBeans = 1024.0 * prob;
+    float gain = (log(nBeans)  / log(4.)) * (40/nBeans);
+
+    p->probability = prob;
+    p->gain = gain;
+}
+
+void        tShaker_setResonance (tShaker *const pm, Lfloat freq, Lfloat radius) {

```

Comment:
I think that getting rid of the `resonator` field entirely, and instead returning the direct output `p->level * tNoise_tick(&p->noise)` on each tick, would make for a better API. This is fine because there's no feedback.

If we let the user handle the filtering/resonance, they can set up and modify as many resonating stages as they like (which is used on almost all of Cook's PhISEM presets), and we don't have to expose a wrapper to `tBiQuad_setResonance` for `tShaker`.